gaiku-common/src/data/chunk.rs

```diff
@@ -5,13 +5,26 @@ pub trait Chunkify {
   fn depth(&self) -> usize;
   fn is_air(&self, x: usize, y: usize, z: usize) -> bool;
   fn get(&self, x: usize, y: usize, z: usize) -> u8;
+  fn get_color(&self, x: usize, y: usize, z: usize) -> Vector4<u8>;
   fn height(&self) -> usize;
   fn position(&self) -> Vector3<f32>;
   fn width(&self) -> usize;
+  fn get_with_color(&self, x: usize, y: usize, z: usize) -> (u8, Vector4<u8>) {
+    (self.get(x, y, z), self.get_color(x, y, z))
+  }
 }
 
 pub trait ChunkifyMut {
   fn set(&mut self, x: usize, y: usize, z: usize, value: u8);
+  fn set_color(&mut self, x: usize, y: usize, z: usize, color: Vector4<u8>);
+  fn set_with_color(&mut self, x: usize, y: usize, z: usize, value: u8, color: Vector4<u8>) {
+    self.set(x, y, z, value);
+    self.set_color(x, y, z, color);
+  }
+}
+
+pub trait ChunkifyNeighboured {
+  fn update_neighbor_data<T: Chunkify + ChunkifyNeighboured + Clone>(&self, _neighbor: &T);
```

Comment:
One thing that is in here but not in #14 is the colors as well as some provided methods with them. Could we add theses to #14 or is there something else going on there with colors?

Reply:
I was thinking in #14 that the colors could be added to a `Texture2d` (`Texturify`), in the examples the texture is created in the `FileFormat` and used in the `Baker`, if you check the voxel baker you will note that there is a type now (u8, u8), in this case the first one is the index in the texture and the second one is the value of the block, so you could use the same methodology maybe?

I don't know how you are generating the colors right now, if you can provide a snippet on how are you getting the colors we can maybe figure out a solution that works for "everyone", because I think most projects might use colors in the prototype stage only and in later stages will want to use textures.

Also right now the only downside (but easily addressable) is that the texture supports 256 colors (because the index for the atlas in the chunk is `u8`).

Reply:
Another solution could be crating another voxel baker (in the 3d crate) called ColoredVoxel or similar, and set the type as `([u8; 4], u8)` maybe? We could move to a common module inside voxel that contains the common code between the Voxel and ColoredVoxel.

We might need minor changes in `Meshify` to support custom data, example:

```rust
impl Meshify for Mesh {
    fn set_attributes(&mut self, attribute: &str, values: Vec<f32>) {
        self.attributes.insert(attribute, values);
    }
   fn get_attributes(&mut self, attribute: &str,) Option<Vec<f32>> { ... }
}
```

We can do it much more flexible by using an enum instead of a `Vec<f32>` that has support for more types like [this](https://github.com/bevyengine/bevy/blob/6a0116e994cb7d9d105cb0897f542701d955db88/crates/bevy_render/src/mesh/mesh.rs#L20).

Reply:
| I don't know how you are generating the colors right now

Currently in the stage of just fixed colors for different things or different LODs so yep prototyping haha. Next step will be to just use the noise function to make shades of browns and greens rather than just solid. Then procedural textures, if I ever get that far :).

If I understand your snippit right you are setting values like like amethyst `MeshBuilder` `.with(<DATA>)`  to add arbitary data?

https://github.com/norman784/gaiku/blob/a2c5ca0b1d3b12a8eda85a71a7b7f9fcd1458b25/gaiku-amethyst/src/mesher.rs#L66-L73

Reply:
The data comes from the texture, the texture is generated with the information of the gox file (this process happens in the FileFormat),

https://github.com/norman784/gaiku/blob/a2c5ca0b1d3b12a8eda85a71a7b7f9fcd1458b25/gaiku-3d/src/formats/gox.rs#L70-L74

In your case will be different, but you can adapt you code to use a different texture for each lot, or maybe have a base texture (the texture is just a byte array containing the color information, each pixel has 4 elements R, G, B, A) with fixed index for each LOD, for example nearest chunk is 0, then 1, 2, etc.

https://github.com/norman784/gaiku/blob/a2c5ca0b1d3b12a8eda85a71a7b7f9fcd1458b25/gaiku-amethyst/src/lib.rs#L159-L165

The idea on the textures needs more work to be finish, so expect some changes around that.